src/nnet3/nnet-general-component.cc

```diff
@@ -880,6 +880,263 @@ void StatisticsPoolingComponent::Backprop(
+                             const CuMatrixBase<BaseFloat> &, //in_value
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &out_deriv,
+                             Component *to_update_in, // may be NULL; may be
+                             // identical to "this" or different.
+                             CuMatrixBase<BaseFloat> *in_deriv) const {
+  const BackpropTruncationComponentPrecomputedIndexes *indexes =
+      dynamic_cast<const BackpropTruncationComponentPrecomputedIndexes*>(
+          indexes_in);
+  KALDI_ASSERT(indexes->zeroing.Dim() == out_deriv.NumRows());
+  // the following statement will do nothing if in_deriv and out_deriv have same
+  // memory.
+  in_deriv->CopyFromMat(out_deriv);
+
+  BackpropTruncationComponent *to_update =
+      dynamic_cast<BackpropTruncationComponent*>(to_update_in);
+
+  // computes clipping_scales
+  BaseFloat clipping_threshold =
+      (clipping_threshold_ <= 0.0 ? 1.0e+10 : clipping_threshold_);
+  // each row in the derivative matrix, which corresponds to one sample in
+  // the mini-batch, is scaled to have a max-norm of clipping_threshold_
+  CuVector<BaseFloat> clipping_scales(in_deriv->NumRows());
+  clipping_scales.AddDiagMat2(pow(clipping_threshold, -2), *in_deriv,
+                              kNoTrans, 0.0);
+  // now clipping_scales contains the squared (norm of each row divided by
+  //  clipping_threshold)
+  int32 num_not_scaled = clipping_scales.ApplyFloor(1.0);
+  // now clipping_scales contains min(1, squared-(norm/clipping_threshold))
+  clipping_scales.ApplyPow(-0.5);
+  // now clipping_scales contains max(1, clipping_threshold/vector_norm)
+  if (to_update != NULL) {
+    to_update->num_clipped_ += (clipping_scales.Dim() - num_not_scaled);
+    to_update->count_ += clipping_scales.Dim();
+  }
+
+  // computes zeroing_scales
+  BaseFloat zeroing_threshold =
+      (zeroing_threshold_ <= 0.0 ? 1.0e+10 : zeroing_threshold_);
+  // zeroing_scales_vec is actually a 1-row matrix.  (the ApplyHeaviside
+  // function isn't defined for vectors).
+  CuMatrix<BaseFloat> zeroing_scales(1, in_deriv->NumRows());
+  CuSubVector<BaseFloat> zeroing_scales_vec(zeroing_scales, 0);
+  zeroing_scales_vec.Set(-pow(zeroing_threshold, 2));
+  // now zeroing_scales_vec contains -(squared zeroing_threshold)
+  zeroing_scales_vec.AddDiagMat2(1.0, *in_deriv, kNoTrans, 1.0);
+  // now zeroing_scales_vec contains squared norm of each row -
+  // squared zeroing_threshold
+  zeroing_scales.ApplyHeaviside();
+  // now the element of zeroing_scales_vec is 1.0 if its corresponding
+  // sample's norm exceeds zero_threshold, and 0.0 otherwise
+  zeroing_scales_vec.MulElements(indexes->zeroing);
+  if (to_update != NULL) {
+    to_update->num_zeroed_ += static_cast<int32>(zeroing_scales_vec.Sum());
+    to_update->count_zeroing_boundaries_ +=
+        static_cast<int32>(indexes->zeroing.Sum());
+  }
+  zeroing_scales_vec.Scale(-1.0);
```

Comment:
Perhaps you could get rid of the need for this by introducing the factor of -1 to indexes->zeroing?